content/blog/2022/zero-units.md

```diff
@@ -0,0 +1,348 @@
+
+```css
+html {
+  /* valid colors */
+  color: hsl(0 60% 80%);
+  color: hsl(0deg 60% 80%);
+
+  /* invalid colors */
+  color: hsl(60 0 80);
+  color: hsl(60deg 0 80%);
+}
+```
+
+The same is true inside the `calc()` function.
+Numbers can be added or subtracted with other numbers,
+and lengths can be added or subtracted with other lengths.
+_It is invalid to add or subtract a number with a length_.
+And that is true _even if the number or length is zero_:
+
+{{ embed.codepen(
+  id='RwMQRpa',
+  title='some zeros need units',
+  user='miriamsuzanne',
+  height=500
+) }}
+
+
+It turns out this is also documented
+[in the specification](https://drafts.csswg.org/css-values-4/#typedef-calc-sum)
+for the `calc()` function:
+
+> Note: Because `<number-token>`s
+> are always interpreted as `<number>`s or `<integer>`s,
+> "unitless 0" `<length>`s aren’t supported in math functions.
+> That is, `width: calc(0 + 5px);` is invalid,
+> because it’s trying to add a `<number>` to a `<length>`,
+> even though both `width: 0;` and `width: 5px;` are valid.
+
+## Fixing the outdent use-case
+
+This is the issue with my `--outdent` code above:
+
+```css
+ol, ul {
+  /* (1em + 0 == 1em) and (1em + -1em == 0) */
+  padding-inline-start: calc(1em + var(--outdent));
+}
+
+li {
+  /* nested lists should not outdent */
+  --outdent: 0;
+}
+```
+
+When `--outdent` is zero (without any units),
+the `calc()` function becomes `calc(0 + 1em)` --
+a `<number>` being added to a `<length>` --
+which is invalid.
+The entire declaration is ignored,
+and no `padding` is applied.
+
+The fix is simple:
+add units to the `--outdent`,
+even when the value is zero:
+
+```css
+li {
+  /* any length units will work here */
+  --outdent: 0px;
+}
+```
+
+And the reason that I keep fixing this same issue
+over and over
+is because I use a linter that doesn't understand the issue.
+That linter runs every time I commit my code,
+and wipes out the units that I've supplied.
+
+## Be cautious with linting & minification
+
+Depending on the linter,
+I can likely turn off that particular 'optimization' --
+and then also file an issue for them to fix.
+That's fine.
+I understand
+that it is not easy to account for every edge case.
+There will always be issues that come up.
+
+But these problems are exacerbated
+by tools like linters & minifiers
+which apply opinionated transformations
+to already-valid code.
+
+I had a similar issue last week,
+with a CSS minifier
+_removing all Cascade Layers from my CSS_.
+In one case,
+the transformation is an over-eager 'best practice'.
+In the other it's an over-eager attempt
+to remove 'unknown syntax'.
+In both cases,
+_I wish linters and minifiers
+would be less eager to transform my code_.
+
+I think we (as an industry)
+tend to adopt rules and 'best practices' very quickly,
+without communicating the caveats clearly,
+and then we fail to update our understanding as things change.
+Tools based on these 'best practice' trends
```

Comment:
Earlier in this article "best practices" was in Title Case, but this is all lowercase. Not sure if there is a reason to switch when it's written here?